File: 207-course-schedule/207-course-schedule.cpp
class Solution {
public:
    bool isCycle(int n, vector<bool>& taken, vector<bool>& visited,  map<int, vector<int>>& graph){
        if (taken[n]){
            return false;
        }
        if(visited[n]){
            return true;
        }
        visited[n] = true;
        for (auto prereq: graph[n]){
            if (isCycle(prereq, taken, visited, graph)){
                return true;
            }
        }
        taken[n] = true;
        visited[n] = false;
        return false;
    }
    
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        map<int, vector<int>> graph;
        vector<bool> visited(numCourses, false);
        vector<bool> taken(numCourses, false);
        
        for (int i= 0; i<prerequisites.size(); ++i){
            graph[prerequisites[i][0]].push_back(prerequisites[i][1]);
        }
        
        for (int i=0; i < numCourses; ++i){
            if (isCycle(i, taken, visited, graph)){
                return false;
            }
        }
        
       return true;
        
        
        
    }
};

File: 207-course-schedule/207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = defaultdict(set)
        
        for course, prereq in prerequisites:
            graph[course].add(prereq)
            
        takencourses = set()
        for i in range(numCourses):
            self.dfs(i, takencourses, graph, set())
        
        return len(takencourses) == numCourses
   
    def dfs(self, n, takencourses, graph, visited):
        if n in takencourses:
            return True 
        if n in visited:
            return False
        visited.add(n)
        for prereq in graph[n]:
            if not self.dfs(prereq, takencourses, graph, visited):
                return False
        takencourses.add(n)
        visited.remove(n)
        return True
        